subsys/net/lib/mqtt/mqtt_encoder.c

```diff
@@ -998,40 +964,189 @@ int publish_encode(const struct mqtt_client *client,
+	if (prop != NULL) {
+		/* The Reason Code and Property Length can be omitted if the
+		 * Reason Code is 0x00 (Success) and there are no Properties.
+		 */
+		if (common_ack_properties_length(prop) == 0 &&
+		    reason_code == 0) {
+			goto out;
+		}
+
+		err_code = pack_uint8(reason_code, buf);
+		if (err_code != 0) {
+			return err_code;
+		}
+
+		err_code = common_ack_properties_encode(prop, buf);
+		if (err_code != 0) {
+			return err_code;
+		}
+	}
+
+out:
+	return mqtt_encode_fixed_header(message_type, start, buf);
+}
+
+int publish_ack_encode(const struct mqtt_client *client,
+		       const struct mqtt_puback_param *param,
 		       struct buf_ctx *buf)
 {
 	const uint8_t message_type =
 		MQTT_MESSAGES_OPTIONS(MQTT_PKT_TYPE_PUBACK, 0, 0, 0);
+	const struct mqtt_common_ack_properties *prop = NULL;
+	uint8_t reason_code = 0;
 
-	return mqtt_message_id_only_enc(message_type, param->message_id, buf);
+#if defined(CONFIG_MQTT_VERSION_5_0)
+	if (mqtt_is_version_5_0(client)) {
+		prop = &param->prop;
+		reason_code = param->reason_code;
+	}
+#endif
+
+	return common_ack_encode(message_type, param->message_id,
+				 reason_code, prop, buf);
 }
 
-int publish_receive_encode(const struct mqtt_pubrec_param *param,
+int publish_receive_encode(const struct mqtt_client *client,
+			   const struct mqtt_pubrec_param *param,
 			   struct buf_ctx *buf)
 {
 	const uint8_t message_type =
 		MQTT_MESSAGES_OPTIONS(MQTT_PKT_TYPE_PUBREC, 0, 0, 0);
+	const struct mqtt_common_ack_properties *prop = NULL;
+	uint8_t reason_code = 0;
 
-	return mqtt_message_id_only_enc(message_type, param->message_id, buf);
+#if defined(CONFIG_MQTT_VERSION_5_0)
+	if (mqtt_is_version_5_0(client)) {
+		prop = &param->prop;
+		reason_code = param->reason_code;
+	}
+#endif
+
+	return common_ack_encode(message_type, param->message_id,
+				 reason_code, prop, buf);
 }
 
-int publish_release_encode(const struct mqtt_pubrel_param *param,
+int publish_release_encode(const struct mqtt_client *client,
+			   const struct mqtt_pubrel_param *param,
 			   struct buf_ctx *buf)
 {
 	const uint8_t message_type =
 		MQTT_MESSAGES_OPTIONS(MQTT_PKT_TYPE_PUBREL, 0, 1, 0);
+	const struct mqtt_common_ack_properties *prop = NULL;
+	uint8_t reason_code = 0;
 
-	return mqtt_message_id_only_enc(message_type, param->message_id, buf);
+#if defined(CONFIG_MQTT_VERSION_5_0)
+	if (mqtt_is_version_5_0(client)) {
+		prop = &param->prop;
+		reason_code = param->reason_code;
+	}
+#endif
+
+	return common_ack_encode(message_type, param->message_id,
+				 reason_code, prop, buf);
 }
 
-int publish_complete_encode(const struct mqtt_pubcomp_param *param,
+int publish_complete_encode(const struct mqtt_client *client,
+			    const struct mqtt_pubcomp_param *param,
 			    struct buf_ctx *buf)
 {
 	const uint8_t message_type =
 		MQTT_MESSAGES_OPTIONS(MQTT_PKT_TYPE_PUBCOMP, 0, 0, 0);
+	const struct mqtt_common_ack_properties *prop = NULL;
+	uint8_t reason_code = 0;
+
+#if defined(CONFIG_MQTT_VERSION_5_0)
```

Comment:
Just wondering if it is possible to add
```
if (!IS_ENABLED(CONFIG_MQTT_VERSION_5_0)) {
	return false;
}
...
```
to `mqtt_is_version_5_0()` so we could avoid these ifdef's?

Reply:
There already is `IS_ENABLED(CONFIG_MQTT_VERSION_5_0)` check inside `mqtt_is_version_5_0()` and I rely on this only when possible. However, some fields within MQTT structs are conditionally compiled out if MQTT 5.0 is disabled, so in a few places I had to use `#if defined(CONFIG_MQTT_VERSION_5_0)` too, otherwise the code wouldn't build for MQTT 3.1.1.

Reply:
Ok, I guessed it was something like that. Never mind then.